### ISSUE TITLE
SLP tests are failing in Travis CI

### ISSUE THREAD

**Issue by christroutner** [Bitcoin-com/rest.bitcoin.com#415]
The scope of this issue is to determine why the SLP route tests are failing in Travis CI and fix them.

Specifically, Gabriel assigned me the task of trying to figure out why this test is failing after the typescript conversion:
https://github.com/Bitcoin-com/rest.bitcoin.com/blob/de3d83d366a34b863ac947ec67712d3b36968904/test/v2/slp.js#L905-L919

## How the test broke
- This is a link to a previous commit in the rest repository, before the typescript refactor, when all the unit tests were working:
https://github.com/Bitcoin-com/rest.bitcoin.com/tree/e15af218ee1eef3e91238a7cde3d2b9aa4306cec

- You can see that the slp.js route was fundamentally structured differently, with slpjs as a direct dependency:
https://github.com/Bitcoin-com/rest.bitcoin.com/blob/e15af218ee1eef3e91238a7cde3d2b9aa4306cec/src/routes/v2/slp.ts#L20-L22

- This is the test in question, that Gabriel asked me to fix:
https://github.com/Bitcoin-com/rest.bitcoin.com/blob/de3d83d366a34b863ac947ec67712d3b36968904/test/v2/slp.js#L906-L910

- Proxyquire is used to create the `slpRouteStub` object to stub out the slpjs library, so that the unit tests do not depend on this external library.
https://github.com/Bitcoin-com/rest.bitcoin.com/blob/de3d83d366a34b863ac947ec67712d3b36968904/test/v2/slp.js#L23

- What the above stub does: When this line (old code) is called, `slp.BitboxNetwork` returns mocked data, instead of live data coming from the slpjs library.
https://github.com/Bitcoin-com/rest.bitcoin.com/blob/e15af218ee1eef3e91238a7cde3d2b9aa4306cec/src/routes/v2/slp.ts#L1274

- The mocked data it returns lives here:
https://github.com/Bitcoin-com/rest.bitcoin.com/blob/de3d83d366a34b863ac947ec67712d3b36968904/test/v2/mocks/slpjs-mocks.js#L13-L38

## Why the test is failing in the `stage` branch
- The slp.ts route in rest has been refactored to remove slpjs as a direct depenency, so the assumptions in the previous test are no longer valid. This test should have broken, and did. This is a good thing. When you have major refactors like this, it is expected to break tests. The test needs to be completely re-written to conform to the new code paths and assumptions.

- Instead of directly depending on slpjs, the slp.ts route now depends only on SLPSDK and pulls out the slpjs library as a depenedency of a depenency. As a result, the same mocking technique will no longer work.

## How to fix the test
- Commenting out the 'unit' test portion, will cause the tests to pass locally, but they fail in Travis CI.
https://github.com/Bitcoin-com/rest.bitcoin.com/blob/de3d83d366a34b863ac947ec67712d3b36968904/test/v2/slp.js#L906-L910

- The error in Travis CI is as follows:
```
{ TypeError [ERR_INVALID_ARG_TYPE]: The "url" argument must be of type string. Received type undefined
	at Url.parse (url.js:154:11)
	at Object.urlParse [as parse] (url.js:148:13)
	at dispatchHttpRequest (/home/travis/build/Bitcoin-com/rest.bitcoin.com/node_modules/axios/lib/adapters/http.js:57:22)
	at new Promise (<anonymous>)
	at httpAdapter (/home/travis/build/Bitcoin-com/rest.bitcoin.com/node_modules/axios/lib/adapters/http.js:18:10)
	at dispatchRequest (/home/travis/build/Bitcoin-com/rest.bitcoin.com/node_modules/axios/lib/core/dispatchRequest.js:59:10)
  [stack]:
   'TypeError [ERR_INVALID_ARG_TYPE]: The "url" argument must be of type string. Received type undefined\n    at Url.parse (url.js:154:11)\n    at Object.urlParse [as parse] (url.js:148:13)\n    at dispatchHttpRequest (/home/travis/build/Bitcoin-com/rest.bitcoin.com/node_modules/axios/lib/adapters/http.js:57:22)\n    at new Promise (<anonymous>)\n    at httpAdapter (/home/travis/build/Bitcoin-com/rest.bitcoin.com/node_modules/axios/lib/adapters/http.js:18:10)\n    at dispatchRequest (/home/travis/build/Bitcoin-com/rest.bitcoin.com/node_modules/axios/lib/core/dispatchRequest.js:59:10)',
  [message]:
   'The "url" argument must be of type string. Received type undefined' }
```

- The above error is coming from this line in the slp route. The failure is originating from the slpjs library:
https://github.com/Bitcoin-com/rest.bitcoin.com/blob/e15af218ee1eef3e91238a7cde3d2b9aa4306cec/src/routes/v2/slp.ts#L1277

- Most likely there is some environment variable that is not set incorrectly.

- I don't see any easy, quick fix to these issues. If the approach we're taking is to no longer mock external dependencies and instead have the only tests be integration tests, with live network calls and live dependencies, then the entire slp test file needs to be refactored with this base assumption.

**Comment by christroutner** [comment]
Created PR #416 to remove unit tests from the slp.js test library. All tests are now passing in Travis CI.

**Comment by christroutner** [comment]
Ok. All tests pass when run locally, but not in Travis. The CI failure persists after removing all references to stubs or mocked data. And the conclusion is the same: the error is originating from either SLP-SDK or slpjs library. There seems to be an environment variable missing.

**Comment by christroutner** [comment]
I thought perhaps it might be that SLPDB is not instantiated for testnet in the SLP route, so I added these lines:

```javascript
let SLP: any = new SLPSDK()
if(process.env.NETWORK==='testnet') {
  SLP = new SLPSDK({restURL: 'https://trest.bitcoin.com/v2/'})
}
```

But that didn't change any behavior.

I think solving this problem will require @jcramer or someone more familiar with the inner workings of slpjs.

**Comment by christroutner** [comment]
Travis CI is failing with this error:

```
{ Error: connect ECONNREFUSED 127.0.0.1:80
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1097:14)
  [stack]:
   'Error: connect ECONNREFUSED 127.0.0.1:80\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1097:14)',
  [message]: 'connect ECONNREFUSED 127.0.0.1:80',
  errno: 'ECONNREFUSED',
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 80,
  config:
   { adapter: [Function],
     transformRequest: [Object],
     transformResponse: [Object],
     timeout: 0,
     xsrfCookieName: 'XSRF-TOKEN',
     xsrfHeaderName: 'X-XSRF-TOKEN',
     maxContentLength: -1,
     validateStatus: [Function],
     headers: [Object],
     method: 'post',
     baseURL: 'undefined',
     auth: [Object],
     data:
      '{"jsonrpc":"1.0","id":"getrawtransaction","method":"getrawtransaction","params":["57b3082a2bf269b3d6f40fee7fb9c664e8256a88ca5ee2697c05b9457822d446",0]}',
     url: 'undefined' },
  request:
   Writable {
     _writableState: [WritableState],
     writable: true,
     _events: [Object],
     _eventsCount: 2,
     _maxListeners: undefined,
     _options: [Object],
     _ended: false,
     _ending: true,
     _redirectCount: 0,
     _redirects: [Array],
     _requestBodyLength: 151,
     _requestBodyBuffers: [Array],
     _onNativeResponse: [Function],
     _currentRequest: [ClientRequest],
     _currentUrl: 'http:undefined' },
```

**Comment by christroutner** [comment]
I was finally able to replicate the errors above locally. 

The error that starts with `ECONNREFUSED 127.0.0.1:80` only happens on the first installation. I had to delete the `rest.bitcoin.com` local repo and reinstall it with dependencies to get it to show up. 

After that first error, the error changes to the one about `The "url" argument must be of type string.`. That error appears instead every time the tests are run, after the first time.

**Comment by christroutner** [comment]
I was able to open up slpjs in the `node_modules` directory and print out `this.BITBOX.restURL` value and confirm that the BITBOX instance that `slpjs.BitboxNetwork.getTransactionDetails()` is using is indeed testnet (trest). So that's no the cause of the errors.

**Comment by christroutner** [comment]
At Gabriels direction, I created a new PR #417 to mock out the unit test by using a different approach to mocking.

**Comment by cgcardona** [comment]
This is resolved in [PR 417](https://github.com/Bitcoin-com/rest.bitcoin.com/pull/417)

**Comment by christroutner** [comment]
This Issue appears to have been fixed in PR #417? Closing for now. We can open a new issue if this reoccurs.
